libdirections-offboard/src/main/java/com/mapbox/navigation/route/offboard/MapboxOffboardRouter.kt

```diff
@@ -20,41 +15,45 @@ class MapboxOffboardRouter(
 ) : Router {
 
     private var navigationRoute: NavigationRoute? = null
+    // For testing only
+    private var isNeedToBuildNavigationRoute = true
+    // For testing only
+    private lateinit var routeRetrieveCallback: RouteRetrieveCallback
 
     override fun getRoute(
         origin: Point,
         waypoints: List<Point>?,
         destination: Point,
         callback: Router.Callback
     ) {
-        val builder = NavigationRoute
-            .builder(context)
-            .accessToken(mapboxToken)
-            .origin(origin)
-            .destination(destination)
-        waypoints?.forEach { builder.addWaypoint(it) }
-        navigationRoute = builder.build()
-        navigationRoute?.getRoute(object : Callback<DirectionsResponse> {
-            override fun onFailure(call: Call<DirectionsResponse>, t: Throwable) {
-                callback.onFailure(t)
-            }
-
-            override fun onResponse(
-                call: Call<DirectionsResponse>,
-                response: Response<DirectionsResponse>
-            ) {
-                val route = response.body()?.routes()
-                if (response.isSuccessful && !route.isNullOrEmpty()) {
-                    callback.onRouteReady(route.map { it.mapToRoute() })
-                } else {
-                    callback.onFailure(NavigationException("Error fetching route"))
-                }
-            }
-        })
+        if (isNeedToBuildNavigationRoute) {
+            val builder = NavigationRoute
+                .builder(context)
+                .accessToken(mapboxToken)
+                .origin(origin)
+                .destination(destination)
+            waypoints?.forEach { builder.addWaypoint(it) }
+            navigationRoute = builder.build()
+        }
+        if (!::routeRetrieveCallback.isInitialized) {
+            routeRetrieveCallback = RouteRetrieveCallback(callback)
+        }
+        navigationRoute?.getRoute(routeRetrieveCallback)
     }
 
     override fun cancel() {
         navigationRoute?.cancelCall()
         navigationRoute = null
     }
+
+    // For testing only
+    internal fun setNavigationRoute(navigationRoute: NavigationRoute) {
```

Comment:
To write clean tests and not pollute the class with test setups we'd need to inject the `NavigationRoute`. Unfortunately, we cannot do it, because the our `libnavigation-core` might not have a reference to the class, if it's been excluded.

In this case, we can override constructors:
```
class MapboxOffboardRouter : Router {

    private val context: Context
    private val mapboxToken: String
    private val routeBuilderProvider: () -> NavigationRoute.Builder

    constructor(context: Context, mapboxToken: String) : this(context, mapboxToken, {
        NavigationRoute
            .builder(context)
            .accessToken(mapboxToken)
    })

    internal constructor(context: Context, mapboxToken: String, routeBuilderProvider: () -> NavigationRoute.Builder) {
        this.context = context
        this.mapboxToken = mapboxToken
        this.routeBuilderProvider = routeBuilderProvider
    }
```

The test code can then invoke the internal constructor (and mock the builder, effectively mocking the route as well), while the production code will wire through it as well. This gives us mostly the same code path, with only a trivial public constructor invocation untested.